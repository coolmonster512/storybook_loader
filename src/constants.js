const appName = 'storybook-config-writer';

export {
  appName, // eslint-disable-line import/prefer-default-export
};
