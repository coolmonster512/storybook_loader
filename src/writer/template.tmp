// template for doT (https://github.com/olado/doT)

function loadStories() {
  
  {{~it.files :value:index}}require('{{=value.relative}}'); // {{=value.full}}
  {{~}}
}

module.exports = {
  loadStories,
};
