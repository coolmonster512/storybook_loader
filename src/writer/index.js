import fs from 'fs';
import path from 'path';
import dot from 'dot';
import { encoding } from '../constants';

dot.templateSettings.strip = false;

function getRelativePaths(fromDir, files) {
  return files.map((file) => {
    let relativePath = path.relative(fromDir, file);

    if (relativePath.substr(0, 2) !== '..' || relativePath.substr(0, 2) !== './') {
      relativePath = `./${relativePath}`;
    }

    return {
      relative: relativePath,
      full: file,
    };
  });
}

function ensureFileDirectoryExists(filePath) {
  const directory = path.dirname(filePath);

  const stats = fs.lstatSync(directory);

  if (!stats.isDirectory()) {
    fs.mkdirSync(directory);
  }
}

export const templateContents = `
// template for doT (https://github.com/olado/doT)

function loadStories() {
  
  {{~it.files :value:index}}require('{{=value.relative}}'); // {{=value.full}}
  {{~}}
}

module.exports = {
  loadStories,
};
`;

export function writeFile(baseDir, files, outputPath) {
  const template = dot.template(templateContents);
  const relativePaths = getRelativePaths(path.dirname(outputPath), files);

  const output = template({ files: relativePaths });

  ensureFileDirectoryExists(outputPath);

  fs.writeFileSync(outputPath, output, { encoding });
}
