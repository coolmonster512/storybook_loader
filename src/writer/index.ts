import { promises as fs } from "fs";
import path from "path";

import {
  getRelativePath,
  ensureFileDirectoryExists
} from "../utils/pathHelper";
import { encoding } from "../constants";

const getRelativePaths = (fromDir: string, files: string[]) => {
  return files
    .map(file => getRelativePath(file, fromDir))
    .concat()
    .sort();
};

const formatter = (
  files: string[],
  frms: (file: string) => string,
  separator: string
) => {
  const formatted = files.map(f => frms(f));
  return formatted.join(separator);
};

const template = (files: string[]) =>
  `// Auto-generated file created by react-native-storybook-loader
// Do not edit.
//
// https://github.com/elderfo/react-native-storybook-loader.git

function loadStories() {
${formatter(files, file => `\trequire('${file}');`, "\n")}
}

const stories = [
${formatter(files, file => `\t'${file}'`, ",\n")}
];

module.exports = {
  loadStories,
  stories,
};
`;

export const writeFile = async (files: string[], outputFile: string) => {
  const relativePaths = getRelativePaths(path.dirname(outputFile), files).map(
    file => file.substring(0, file.lastIndexOf("."))
  ); // strip file extensions

  const output = template(relativePaths);

  await ensureFileDirectoryExists(outputFile);

  await fs.writeFile(outputFile, output, { encoding });
};
