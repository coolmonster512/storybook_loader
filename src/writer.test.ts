import mock from "mock-fs";
import path from "path";
import { promises as fs } from "fs";
import { writeFile } from "./writer";
import { encoding } from "./constants";

const baseDir = path.resolve(__dirname, "../../");
const outputPath = path.resolve(baseDir, "output.js");
const moduleDir = path.resolve(baseDir);
const configFile = path.resolve(moduleDir, "./module.config.json");

beforeEach(() => {
  mock({
    [outputPath]: "",
    [configFile]: ""
  });
});

afterEach(() => {
  mock.restore();
});

describe("writeFile", () => {
  test("generate story loader", async () => {
    const files = [
      path.resolve(__dirname, "../file1.js"),
      path.resolve(__dirname, "../sub/file2.js"),
      path.resolve(__dirname, "../../parent/file3.js"),
      path.resolve(__dirname, "./sub/file4.js"),
      path.resolve(__dirname, ".\\sub\\sub\\file5.js")
    ];
    await writeFile(files, outputPath);

    const contents = await fs.readFile(outputPath, encoding);

    expect(contents).toMatchSnapshot();
  });

  test("should generate dummy story loader if no stories were found", async () => {
    const files: string[] = [];
    writeFile(files, outputPath);

    const contents = await fs.readFile(outputPath, encoding);

    expect(contents).toMatchSnapshot();
  });
});
