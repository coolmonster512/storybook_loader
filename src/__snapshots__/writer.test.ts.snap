// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`writeFile generate story loader 1`] = `
"// Auto-generated file created by react-native-storybook-loader
// Do not edit.
//
// https://github.com/elderfo/react-native-storybook-loader.git

function loadStories() {
	require('./parent/file3');
	require('./react-native-storybook-loader/file1');
	require('./react-native-storybook-loader/src/sub/file4');
	require('./react-native-storybook-loader/src/sub/sub/file5');
	require('./react-native-storybook-loader/sub/file2');
}

const stories = [
	'./parent/file3',
	'./react-native-storybook-loader/file1',
	'./react-native-storybook-loader/src/sub/file4',
	'./react-native-storybook-loader/src/sub/sub/file5',
	'./react-native-storybook-loader/sub/file2'
];

module.exports = {
  loadStories,
  stories,
};
"
`;

exports[`writeFile should generate dummy story loader if no stories were found 1`] = `
"// Auto-generated file created by react-native-storybook-loader
// Do not edit.
//
// https://github.com/elderfo/react-native-storybook-loader.git

function loadStories() {

}

const stories = [

];

module.exports = {
  loadStories,
  stories,
};
"
`;
